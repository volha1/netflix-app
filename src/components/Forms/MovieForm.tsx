import React, { ReactElement, useState, memo, useCallback } from 'react';
import * as Yup from 'yup';
import classNames from 'classnames';
import { Formik, Form, Field } from 'formik';
import CloseBtn from '../CloseBtn/index';
import { isValidUrl } from '../../helpers/utils';
import { genres } from '../../helpers/constants';
import './style.scss';
import Movie from '../../types/Movie';

type MovieFormProps = {
  actionText: string;
  onCloseMovieForm: () => void;
  onSubmit: (movieItem: Movie) => void;
  movie: Movie;
};

const MovieSchema = Yup.object().shape({
  title: Yup.string().required('Required field'),
  voteAverage: Yup.number()
    .typeError('Please, enter number')
    .required('Required field')
    .min(1, 'Please, enter number more or equal to 1')
    .max(10, 'Please, enter number less or equal to 10'),
  imgPath: Yup.string()
    .required('Required field')
    .trim()
    .test('test-url', 'Invalid url', (value) => {
      return isValidUrl(value);
    }),
  runtime: Yup.number()
    .typeError('Please, enter number')
    .required('Required field')
    .min(1, 'Please, enter number more or equal to 1'),
  genres: Yup.array().min(1, 'This field is required'),
  overview: Yup.string().required('Required field'),
});

const MovieForm = ({ actionText, onCloseMovieForm, onSubmit, movie = {} }: MovieFormProps): ReactElement => {
  const [genresShown, setGenresShown] = useState(false);

  const showGenres = useCallback((): void => {
    setGenresShown(!genresShown);
  }, [genresShown]);

  return (
    <Formik
      initialValues={movie}
      enableReinitialize
      validationSchema={MovieSchema}
      onSubmit={(values, { resetForm }): void => {
        onSubmit(values);
        resetForm();
      }}
    >
      {({ errors, touched }): ReactElement => {
        return (
          <Form className="movie-form">
            <CloseBtn onClose={onCloseMovieForm} />
            <h1 className="title">{actionText} movie</h1>
            <div className="row">
              <label>
                Title
                <br />
                <Field
                  name="title"
                  className={classNames('left-field', { 'error-field': errors.title && touched.title })}
                />
                <br />
                {errors.title && touched.title && <span>{errors.title}</span>}
              </label>

              <label>
                Release date
                <br />
                <Field type="date" name="releaseDate" className="right-field" />
              </label>
            </div>

            <div className="row">
              <label>
                Movie url
                <br />
                <Field
                  type="url"
                  name="imgPath"
                  className={classNames('left-field', { 'error-field': errors.imgPath && touched.imgPath })}
                  placeholder="https://"
                />
                <br />
                {errors.imgPath && touched.imgPath && <span>{errors.imgPath}</span>}
              </label>
              <label>
                Rating
                <br />
                <Field
                  name="voteAverage"
                  className={classNames('right-field', { 'error-field': errors.voteAverage && touched.voteAverage })}
                  placeholder="7.8"
                />
                <br />
                {errors.voteAverage && touched.voteAverage && <span>{errors.voteAverage}</span>}
              </label>
            </div>
            <div className="row">
              <div className="left-field genres-select">
                <div className="select-title" onClick={showGenres}>
                  <label>
                    Genre
                    <br />
                    <select className={classNames('left-field', { 'error-field': errors.genres && touched.genres })}>
                      <option>Select genre</option>
                    </select>
                  </label>
                  <br />
                  {errors.genres && touched.genres && <span>{errors.genres}</span>}
                  <div className="overSelect" />
                </div>
                {genresShown && (
                  <div className="genres" role="group">
                    {genres.map((genre) => {
                      return (
                        <label key={genre}>
                          <Field type="checkbox" name="genres" value={genre} />
                          {genre}
                        </label>
                      );
                    })}
                  </div>
                )}
              </div>
              <label>
                Runtime
                <br />
                <Field
                  name="runtime"
                  className={classNames('right-field', { 'error-field': errors.runtime && touched.runtime })}
                  placeholder="minutes"
                />
                <br />
                {errors.runtime && touched.runtime && <span>{errors.runtime}</span>}
              </label>
            </div>
            <div className="row">
              <label>
                Overview
                <br />
                <Field
                  as="textarea"
                  name="overview"
                  className={classNames({ 'error-field': errors.overview && touched.overview })}
                  rows={8}
                  cols={87}
                  placeholder="Movie description"
                />
                <br />
                {errors.overview && touched.overview && <span>{errors.overview}</span>}
              </label>
            </div>
            <div className="btn-wrapper">
              <button type="reset" className="btn">
                Reset
              </button>
              <button type="submit" className="btn">
                Submit
              </button>
            </div>
          </Form>
        );
      }}
    </Formik>
  );
};

export default memo(MovieForm);
