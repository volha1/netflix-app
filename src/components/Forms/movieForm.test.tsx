import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import MovieForm from './MovieForm';

describe('MovieForm test', () => {
  const actionText = '';
  const onCloseMovieForm = jest.fn();
  const onSubmit = jest.fn();
  const movie = {
    id: '',
    title: '',
    voteAverage: '',
    releaseDate: '',
    imgPath: '',
    genres: [],
    overview: '',
    runtime: 0,
  };

  it('MovieForm snapshot', () => {
    const component = renderer.create(
      <MovieForm actionText={actionText} onCloseMovieForm={onCloseMovieForm} onSubmit={onSubmit} movie={movie} />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('onSubmit works', async () => {
    render(<MovieForm actionText={actionText} onCloseMovieForm={onCloseMovieForm} onSubmit={onSubmit} movie={movie} />);

    const user = userEvent.setup();

    await user.type(screen.getByLabelText(/title/i), 'title');
    await user.type(screen.getByLabelText(/movie url/i), 'https://test.com');
    await user.type(screen.getByLabelText(/rating/i), '8');
    await user.type(screen.getByLabelText(/runtime/i), '120');
    await user.type(screen.getByLabelText(/overview/i), 'overview');
    await user.click(screen.getByTestId('select-title'));
    await user.click(screen.getByTestId('horror'));
    await user.click(screen.getByRole('button', { name: /submit/i }));

    expect(onSubmit).toHaveBeenCalled();
    await waitFor(() => {
      return expect(onSubmit).toHaveBeenCalledWith({
        id: '',
        title: 'title',
        imgPath: 'https://test.com',
        genres: ['horror'],
        voteAverage: '8',
        runtime: '0120',
        overview: 'overview',
        releaseDate: '',
      });
    });
  });
});
