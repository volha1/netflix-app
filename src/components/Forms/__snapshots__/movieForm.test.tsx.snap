// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm test MovieForm snapshot 1`] = `
<form
  action="#"
  className="movie-form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="close-btn"
    data-testid="close-btn"
    onClick={[MockFunction]}
  />
  <h1
    className="title"
  >
     movie
  </h1>
  <div
    className="row"
  >
    <label>
      Title
      <br />
      <input
        className="left-field"
        name="title"
        onBlur={[Function]}
        onChange={[Function]}
        value=""
      />
      <br />
    </label>
    <label
      htmlFor="release-date-field"
    >
      Release date
      <br />
      <input
        className="right-field"
        id="release-date-field"
        name="releaseDate"
        onBlur={[Function]}
        onChange={[Function]}
        type="date"
        value=""
      />
    </label>
  </div>
  <div
    className="row"
  >
    <label>
      Movie url
      <br />
      <input
        className="left-field"
        name="imgPath"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="https://"
        type="url"
        value=""
      />
      <br />
    </label>
    <label>
      Rating
      <br />
      <input
        className="right-field"
        name="voteAverage"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="7.8"
        value=""
      />
      <br />
    </label>
  </div>
  <div
    className="row"
  >
    <div
      className="left-field genres-select"
    >
      <div
        className="select-title"
        data-testid="select-title"
        onClick={[Function]}
      >
        <label>
          Genre
          <br />
          <select
            className="left-field"
          >
            <option>
              Select genre
            </option>
          </select>
        </label>
        <br />
        <div
          className="overSelect"
        />
      </div>
    </div>
    <label>
      Runtime
      <br />
      <input
        className="right-field"
        name="runtime"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="minutes"
        value={0}
      />
      <br />
    </label>
  </div>
  <div
    className="row"
  >
    <label>
      Overview
      <br />
      <textarea
        className=""
        cols={87}
        name="overview"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Movie description"
        rows={8}
        value=""
      />
      <br />
    </label>
  </div>
  <div
    className="btn-wrapper"
  >
    <button
      className="btn"
      type="reset"
    >
      Reset
    </button>
    <button
      className="btn"
      data-testid="submit-btn"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
